File: EnemyAI.cpp
#include "EnemyAI.h"

#include <algorithm> // std::clamp
#include <iostream>

// 内部调试输出宏，对齐项目 LOG_INFO 风格
#define AI_LOG(msg) std::cout << "[EnemyAI] " << msg << "\n"

// ============================================================
// Update — 帧更新主入口
// ============================================================
void EnemyAI::Update(float dt) {
    if (m_isDead) return;

    UpdateDetectionValue(dt);
    EvaluateState();
    ExecuteStateAction(dt);
    UpdateHacking(dt);
}

// ============================================================
// UpdateDetectionValue
// ============================================================
void EnemyAI::UpdateDetectionValue(float dt) {
    if (m_isSpotted) {
        m_detectionValue += DETECTION_INCREASE * dt;
    } else {
        m_detectionValue -= DETECTION_DECREASE * dt;
    }
    m_detectionValue = std::clamp(m_detectionValue, 0.0f, 100.0f);
}

// ============================================================
// EvaluateState
// ============================================================
void EnemyAI::EvaluateState() {
    m_previousState = m_currentState;

    if (m_detectionValue <= 15.0f) {
        m_currentState = EnemyState::Safe;
    } else if (m_detectionValue <= 30.0f) {
        m_currentState = EnemyState::Search;
    } else if (m_detectionValue <= 50.0f) {
        m_currentState = EnemyState::Alert;
    } else {
        m_currentState = EnemyState::Hunt;
    }

    m_stateChanged = (m_currentState != m_previousState);
    if (m_stateChanged) {
        const char* name = nullptr;
        switch (m_currentState) {
            case EnemyState::Safe:   name = "Safe";   break;
            case EnemyState::Search: name = "Search"; break;
            case EnemyState::Alert:  name = "Alert";  break;
            case EnemyState::Hunt:   name = "Hunt";   break;
        }
        AI_LOG("State -> " << name << " | detection=" << m_detectionValue);
    }
}

// ============================================================
// ExecuteStateAction
// ============================================================
void EnemyAI::ExecuteStateAction(float dt) {
    switch (m_currentState) {
        case EnemyState::Safe:   HandleSafeState();   break;
        case EnemyState::Search: HandleSearchState(); break;
        case EnemyState::Alert:  HandleAlertState();  break;
        case EnemyState::Hunt:   HandleHuntState(dt); break;
    }
    m_stateChanged = false; // 消费完毕
}

// ============================================================
// HandleSafeState — 原指令：MoveCommand::Patrol
// ============================================================
void EnemyAI::HandleSafeState() {
    m_huntLockTimer = 0.0f;

    if (m_stateChanged) {
        AI_LOG("Action: Patrol");
    }
}

// ============================================================
// HandleSearchState — 原指令：MoveCommand::Stop
// ============================================================
void EnemyAI::HandleSearchState() {
    m_huntLockTimer = 0.0f;

    if (m_stateChanged) {
        AI_LOG("Action: Stop — scanning surroundings");
    }
}

// ============================================================
// HandleAlertState — 原指令：MoveCommand::MoveToPosition 或 SweepRotate
// ============================================================
void EnemyAI::HandleAlertState() {
    m_huntLockTimer = 0.0f;

    if (m_stateChanged) {
        if (m_isSpotted) {
            AI_LOG("Action: MoveToPosition — moving towards player");
        } else {
            AI_LOG("Action: MoveToPosition — moving to last known position");
        }
    }
}

// ============================================================
// HandleHuntState — 原指令：MoveCommand::ChasePlayer（5 秒锁定期）
// ============================================================
void EnemyAI::HandleHuntState(float dt) {
    if (m_huntLockTimer > 0.0f) {
        m_huntLockTimer -= dt;

        if (m_stateChanged) {
            AI_LOG("Action: ChasePlayer — hunt lock active (" << HUNT_LOCK_DURATION << "s)");
        }
    } else {
        if (m_detectionValue > 50.0f) {
            m_huntLockTimer = HUNT_LOCK_DURATION;
            AI_LOG("Action: ChasePlayer — hunt lock reset");
        }
        // 否则 detectionValue 已降到阈值以下，EvaluateState 自动降级状态
    }
}

// ============================================================
// UpdateHacking
// ============================================================
void EnemyAI::UpdateHacking(float dt) {
    // 不在 Safe 状态时立即中断 Hack
    if (m_currentState != EnemyState::Safe) {
        if (m_inputEnabled || m_isHacked || m_hackProgress > 0.0f) {
            m_inputEnabled = false;
            m_hackProgress = 0.0f;
            m_isHacked     = false;
            AI_LOG("Hack interrupted — enemy no longer in Safe state");
        }
        return;
    }

    if (m_inputEnabled && m_hackKeyPressed) {
        if (m_hackProgress < 100.0f) {
            m_hackProgress += HACK_SPEED * dt;
        }

        if (m_hackProgress >= 100.0f && !m_isHacked) {
            m_hackProgress = 100.0f;
            m_isHacked     = true;
            AI_LOG("Hack complete — Execute is now available");
        }
    }
}

// ============================================================
// OnPlayerEnterKillzone
// ============================================================
void EnemyAI::OnPlayerEnterKillzone() {
    m_inKillzone = true;
    AI_LOG("Player Killzone entered enemy range");

    if (m_currentState == EnemyState::Safe) {
        m_inputEnabled = true;
        AI_LOG("Safe state confirmed — hack input enabled");
    } else {
        const char* name = nullptr;
        switch (m_currentState) {
            case EnemyState::Search: name = "Search"; break;
            case EnemyState::Alert:  name = "Alert";  break;
            case EnemyState::Hunt:   name = "Hunt";   break;
            default:                 name = "Unknown"; break;
        }
        AI_LOG("Not Safe (" << name << ") — hack input disabled");
    }
}

// ============================================================
// OnPlayerExitKillzone
// ============================================================
void EnemyAI::OnPlayerExitKillzone() {
    m_inKillzone   = false;
    m_inputEnabled = false;
    m_hackProgress = 0.0f;
    m_isHacked     = false;
    AI_LOG("Player left killzone — hack progress reset");
}

// ============================================================
// OnPlayerSpotted
// ============================================================
void EnemyAI::OnPlayerSpotted() {
    m_isSpotted = true;
    AI_LOG("Player spotted");
}

// ============================================================
// OnPlayerLost
// ============================================================
void EnemyAI::OnPlayerLost() {
    m_isSpotted = false;
    AI_LOG("Player lost");
}

// ============================================================
// TryExecute
// ============================================================
void EnemyAI::TryExecute() {
    if (m_isHacked && m_inputEnabled) {
        m_isDead = true;
        AI_LOG("Enemy executed and killed");
    }
}

File: enemyAI.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class enemyAI : MonoBehaviour
{
    // ========== 原有字段（保留） ==========
    public float detectionValue = 0.0f;
    public float detectionValueIncrease = 10.0f;
    public float detectionValueDecrease = 5.0f;
    private bool isSpotted = false;

    public bool HackEnble = false;
    public float progress = 0.0f;
    public float progressIncreaseValue = 25f;
    public bool IsHacked = false;
    private bool enbleInput = false;

    private bool Hackable = true;
    private GameObject _targetEnemy;


    private PlayerControls _controls;

[Header("处决设置")]
public GameObject spawnOnDeathPrefab; // 拖入你想要在处决后生成的预制体

    // ========== 移动指令输出系统 ==========
    public enum MoveCommand
    {
        Patrol,         // 巡逻
        Stop,           // 停止
        MoveToPosition, // 移动到指定位置
        ChasePlayer,    // 追击玩家
        SweepRotate     // 原地旋转 360°
    }

    public MoveCommand command = MoveCommand.Patrol;
    public Vector3 commandPosition = Vector3.zero;

    // ========== 状态机 ==========
    public enum EnemyState
    {
        Safe,    // ≤15
        Search,  // 16-30
        Alert,   // 31-50
        Hunt     // >50
    }

    private EnemyState currentState = EnemyState.Safe;
    private EnemyState previousState = EnemyState.Safe;

    // ========== 玩家追踪 ==========
    private Transform playerTransform;
    private Vector3 lastKnownPosition = Vector3.zero;

    // ========== Hunt 锁定期 ==========
    private float huntLockTimer = 0.0f;
    public float huntLockDuration = 5.0f;

    // ========== 距离判定 ==========
    private const float ARRIVAL_DISTANCE = 0.5f;

    // ========== 引用 enemyMovements ==========
    private enemyMovements movementController;

    void Start()
    {
        movementController = GetComponent<enemyMovements>();
        if (movementController == null)
        {
            Debug.LogError("enemyMovements not found on " + gameObject.name);
        }
    }

    void Awake()
    {
        _controls = new PlayerControls();

        // 动作名是 .Hack 和 .Execute
        _controls.player.Hack.started += ctx => HackEnble = true;
        _controls.player.Hack.canceled += ctx => HackEnble = false;

        _controls.player.Execute.performed += ctx => execute();

    }

    void Update()
    {
        // 1. 感知更新
        UpdatePerception();

        // 2. detectionValue 更新
        UpdateDetectionValue();

        // 3. 状态评估
        EvaluateState();

        // 4. 执行动作（输出指令）
        ExecuteStateAction();

        //击杀功能
        Hacking();
    }

    // ========== 1. 感知更新 ==========
    private void UpdatePerception()
    {
        if (isSpotted && playerTransform != null)
        {
            lastKnownPosition = playerTransform.position;

            // 更新 enemyMovements.target
            if (movementController != null)
            {
                movementController.target = playerTransform;
            }
        }
    }

    // ========== 2. detectionValue 更新 ==========
    private void UpdateDetectionValue()
    {
        if (isSpotted)
        {
            detectionValue += detectionValueIncrease * Time.deltaTime;
        }
        else
        {
            detectionValue -= detectionValueDecrease * Time.deltaTime;
        }

        // Clamp 到 [0, 100]
        detectionValue = Mathf.Clamp(detectionValue, 0.0f, 100.0f);
    }

    // ========== 3. 状态评估 ==========
    private void EvaluateState()
    {
        previousState = currentState;

        if (detectionValue <= 15)
        {
            currentState = EnemyState.Safe;
        }
        else if (detectionValue > 15 && detectionValue <= 30)
        {
            currentState = EnemyState.Search;
        }
        else if (detectionValue > 30 && detectionValue <= 50)
        {
            currentState = EnemyState.Alert;
        }
        else if (detectionValue > 50)
        {
            currentState = EnemyState.Hunt;
        }

        // 仅在状态切换时打印
        if (currentState != previousState)
        {
            Debug.Log("State -> " + currentState + " | detection=" + detectionValue.ToString("F1"));
        }
    }

    // ========== 4. 执行动作（输出指令） ==========
    private void ExecuteStateAction()
    {
        switch (currentState)
        {
            case EnemyState.Safe:
                HandleSafeState();
                break;
            case EnemyState.Search:
                HandleSearchState();
                break;
            case EnemyState.Alert:
                HandleAlertState();
                break;
            case EnemyState.Hunt:
                HandleHuntState();
                break;
        }
    }

    // ========== Safe 状态：巡逻 ==========
    private void HandleSafeState()
    {
        huntLockTimer = 0.0f;
        command = MoveCommand.Patrol;
    }

    // ========== Search 状态：固定不动 ==========
    private void HandleSearchState()
    {
        huntLockTimer = 0.0f;
        command = MoveCommand.Stop;
    }

    // ========== Alert 状态：追击或 Sweep ==========
    private void HandleAlertState()
    {
        huntLockTimer = 0.0f;

        // 检查是否到达 lastKnownPosition
        float distanceToLastKnown = Vector3.Distance(transform.position, lastKnownPosition);

        if (distanceToLastKnown <= ARRIVAL_DISTANCE)
        {
            // 到达：发出 Sweep 指令
            command = MoveCommand.SweepRotate;
        }
        else
        {
            // 未到达：移动到位置
            if (isSpotted && playerTransform != null)
            {
                // spotted：追 player.position
                commandPosition = playerTransform.position;
            }
            else
            {
                // not spotted：追 lastKnownPosition
                commandPosition = lastKnownPosition;
            }
            command = MoveCommand.MoveToPosition;
        }
    }

    // ========== Hunt 状态：5 秒锁定期 ==========
    private void HandleHuntState()
    {
        if (huntLockTimer > 0)
        {
            // 锁定期内：无条件追击玩家
            huntLockTimer -= Time.deltaTime;
            command = MoveCommand.ChasePlayer;
        }
        else
        {
            // 锁定期结束
            if (detectionValue > 50)
            {
                // 仍在 Hunt 阈值：重置锁定期
                huntLockTimer = huntLockDuration;
                command = MoveCommand.ChasePlayer;
            }
            // 否则按阈值降级（由 EvaluateState 自动处理）
        }
    }

    // ========== 触发器检测 ==========
    void OnTriggerEnter(Collider other)
    {

        if (other.CompareTag("Player"))
        {
            // 1. 优先检查：如果碰到的是名为 "Killzone" 的子触发器
            if (other.gameObject.name == "Killzone")
            {
                // [你的需求] 当 Killzone 与敌人重叠时输出 Debug
                Debug.Log("系统：检测到玩家的 [Killzone] 触发器已进入敌人碰撞范围！");
                playerTransform = other.transform.root; // 锁定玩家根对象坐标

                // 只有敌人处于 Safe 状态才允许获取暗杀输入权限
                if (currentState == EnemyState.Safe)
                {
                    enbleInput = true;
                    Debug.Log("safe,hackable");
                }
                else
                {
                    Debug.Log("Not safe(" + currentState + "),cant hack");
                }
            }
            else 
        {
            // 如果不是 Killzone 进入（即玩家身体接触），则触发正常的“被发现”逻辑
            isSpotted = true;
            playerTransform = other.transform;
            lastKnownPosition = playerTransform.position;

            if (movementController != null)
            {
                movementController.target = playerTransform;
            }
            Debug.Log("spotted");
        }


        }
        

    }

    void OnTriggerExit(Collider other)
    {
        if (other.CompareTag("Player"))
        {
            if (other.gameObject.name == "Killzone")
            {
                enbleInput = false;
                HackEnble = false;
                progress = 0;
                IsHacked = false;
                Debug.Log("not in rang , reset");
            }
            else
            {
                isSpotted = false;
            }
        }
    }

    // ========== Debug 可视化 ==========
    void OnDrawGizmosSelected()
    {
        // 绘制 lastKnownPosition
        Gizmos.color = Color.red;
        Gizmos.DrawWireSphere(lastKnownPosition, 0.5f);

        // 绘制当前指令目标位置
        if (command == MoveCommand.MoveToPosition)
        {
            Gizmos.color = Color.yellow;
            Gizmos.DrawLine(transform.position, commandPosition);
            Gizmos.DrawWireSphere(commandPosition, 0.3f);
        }
    }

    void Hacking()
    {
        // 实时检查：如果敌人不再安全，强制中断 Hack
        if (currentState != EnemyState.Safe)
        {
            enbleInput = false;
            HackEnble = false;
            progress = 0;
            IsHacked = false;
            return;
        }

        // 只有在范围内且按住 Hack 键时增加进度（删除了重复累加逻辑）
        if (enbleInput && HackEnble)
        {
            if (progress < 100)
            {
                progress += progressIncreaseValue * Time.deltaTime;
            }

            if (progress >= 100 && !IsHacked)
            {
                IsHacked = true;
                Debug.Log("Hack 完成，可以 Execute");
            }
        }
    }

    void execute()
    {
        // 如果进度满了，销毁自己
        if (IsHacked && enbleInput)
        {
            Debug.Log("killed");
            // [新增功能]：在当前敌人位置生成指定的预制体
        if (spawnOnDeathPrefab != null)
        {
            // 在敌人的位置 (transform.position) 和 旋转 (transform.rotation) 创建预制体
            Instantiate(spawnOnDeathPrefab, transform.position, transform.rotation);
        }
            Destroy(this.gameObject);
        }
    }
    // 4. Input System 生命周期管理 (容易漏掉)
    private void OnEnable() => _controls.Enable();  // 【必须添加】
    private void OnDisable() => _controls.Disable(); // 【必须添加】
}
